File: ts/test/errs.test.js
/* Copyright (c) 2026 Richard Rodger, MIT License */
'use strict'

// Per-parse error list: ctx.errs. Every TabnasError records itself on
// its Context's errs list at construction, so in today's fail-fast
// mode the thrown error is also the last (and only) entry. This is the
// groundwork for multi-error collection (ts/doc/lsp-feasibility.md):
// a recovery mode appends here and continues instead of throwing.
//
// Contract pinned here:
//   1. flags-off behaviour is unchanged — errors still throw;
//   2. the thrown error IS the recorded error (same object);
//   3. errs is per-parse — a fresh parse starts empty;
//   4. a clean parse records nothing.

const { describe, it } = require('node:test')
const assert = require('node:assert')

const { Tabnas, TabnasError } = require('..')
const { json } = require('../dist-test/json-plugin')

describe('errs', () => {
  it('records the thrown error on ctx.errs', () => {
    const tn = new Tabnas({ plugins: [json] })
    try {
      tn.parse('{"a":1,}')
      assert.fail('parse should fail')
    } catch (e) {
      assert.ok(e instanceof TabnasError)
      const ctx = e.internal.ctx
      assert.ok(Array.isArray(ctx.errs))
      assert.equal(ctx.errs.length, 1)
      assert.equal(ctx.errs[0], e)
    }
  })

  it('starts empty on every parse', () => {
    const tn = new Tabnas({ plugins: [json] })
    let first
    try {
      tn.parse('[1,')
      assert.fail('parse should fail')
    } catch (e) {
      first = e
    }
    try {
      tn.parse('{"b":')
      assert.fail('parse should fail')
    } catch (e) {
      // Second parse has its own Context and its own errs list: only
      // its own error, and not the first parse's.
      const errs = e.internal.ctx.errs
      assert.equal(errs.length, 1)
      assert.equal(errs[0], e)
      assert.notEqual(errs[0], first)
      assert.notEqual(e.internal.ctx, first.internal.ctx)
    }
  })

  it('records nothing on a clean parse', () => {
    const tn = new Tabnas({ plugins: [json] })
    let seen
    tn.sub({
      rule: (rule, ctx) => {
        seen = ctx
      },
    })
    const out = tn.parse('{"a":[1,2]}')
    // JSON round-trip: the json plugin builds null-prototype objects.
    assert.equal(JSON.stringify(out), '{"a":[1,2]}')
    assert.ok(seen, 'rule sub captured the parse context')
    assert.deepStrictEqual(seen.errs, [])
  })

  it('never throws from recording on a degenerate context', () => {
    // Plugin code can construct TabnasError with unusual ctx values —
    // construction must not depend on ctx.errs being present.
    const err = new TabnasError(
      'unexpected',
      {},
      { isToken: true, tin: 1, rI: 1, cI: 1, sI: 0, len: 0 },
      {},
      {},
    )
    assert.ok(err instanceof TabnasError)
  })
})

File: go/errs_test.go
/* Copyright (c) 2026 Richard Rodger, MIT License */

package tabnas

import "testing"

// Per-parse error list: ctx.Errs, the Go mirror of TS ctx.errs
// (ts/test/errs.test.js). Errors record themselves at their
// CONSTRUCTION site, so the error Parse returns is also the last
// recorded entry. Groundwork for opt-in multi-error recovery.
//
// Pinned here, matching the TS suite case for case:
//  1. fail-fast behaviour is unchanged — Parse still returns an error;
//  2. the returned error IS the recorded error (same pointer);
//  3. Errs is per-parse — a fresh parse starts empty;
//  4. a clean parse records nothing;
//  5. recording never panics on a context-less lexer.

// ctxOf grabs the parse Context via a rule subscriber, which is the
// only handle a caller has on it after Parse returns.
func ctxOf(j *Tabnas) func() *Context {
	var seen *Context
	j.Sub(nil, func(r *Rule, ctx *Context) { seen = ctx })
	return func() *Context { return seen }
}

func TestErrsRecordsReturnedError(t *testing.T) {
	j := makeJSON()
	get := ctxOf(j)

	_, err := j.Parse(`{"a":1,}`)
	if err == nil {
		t.Fatal("parse should fail")
	}
	je, ok := err.(*TabnasError)
	if !ok {
		t.Fatalf("want *TabnasError, got %T", err)
	}
	ctx := get()
	if ctx == nil {
		t.Fatal("no context captured")
	}
	if 1 != len(ctx.Errs) {
		t.Fatalf("want 1 recorded error, got %d", len(ctx.Errs))
	}
	if ctx.Errs[0] != je {
		t.Fatal("returned error is not the recorded error")
	}
}

func TestErrsIsPerParse(t *testing.T) {
	j := makeJSON()
	get := ctxOf(j)

	if _, err := j.Parse(`[1,`); err == nil {
		t.Fatal("parse should fail")
	}
	first := get()

	if _, err := j.Parse(`{"b":`); err == nil {
		t.Fatal("parse should fail")
	}
	second := get()

	if first == second {
		t.Fatal("second parse reused the first parse's context")
	}
	if 1 != len(second.Errs) {
		t.Fatalf("second parse: want 1 recorded error, got %d", len(second.Errs))
	}
}

func TestErrsEmptyOnCleanParse(t *testing.T) {
	j := makeJSON()
	get := ctxOf(j)

	if _, err := j.Parse(`{"a":[1,2]}`); err != nil {
		t.Fatalf("clean parse failed: %v", err)
	}
	ctx := get()
	if ctx == nil {
		t.Fatal("no context captured")
	}
	if 0 != len(ctx.Errs) {
		t.Fatalf("clean parse recorded %d errors", len(ctx.Errs))
	}
}

func TestErrsRecordsLexerErrors(t *testing.T) {
	// An unterminated string is raised by the lexer, not the parser —
	// a different construction site, which must record too.
	j := makeJSON()
	get := ctxOf(j)

	_, err := j.Parse(`{"a":"unterminated`)
	if err == nil {
		t.Fatal("parse should fail")
	}
	ctx := get()
	if ctx == nil || 0 == len(ctx.Errs) {
		t.Fatal("lexer error was not recorded")
	}
	if ctx.Errs[len(ctx.Errs)-1] != err {
		t.Fatal("returned lexer error is not the last recorded error")
	}
}

func TestRecordErrIsNilSafe(t *testing.T) {
	// A Lex built without a Context (NewLex) has no list to record
	// into; recording must never be the thing that breaks a parse.
	var ctx *Context
	je := &TabnasError{Code: "unexpected"}
	if got := ctx.recordErr(je); got != je {
		t.Fatal("recordErr on a nil context must return its argument")
	}
	real := &Context{}
	if got := real.recordErr(nil); got != nil {
		t.Fatal("recordErr(nil) must return nil")
	}
	if 0 != len(real.Errs) {
		t.Fatal("recordErr(nil) must not append")
	}
}
